### PR TITLE
Add calc_fdr_hurdle, a bootstrap hurdle for the multiplicity of a strategy search

### PR DESCRIPTION
**Pull request by ipezygj**
Adds `calc_fdr_hurdle`: the t-statistic hurdle a strategy search implies at a chosen false discovery rate, as a bootstrap plug-in estimate.

It sits next to `calc_deflated_sharpe_ratio` and answers the neighbouring question. The deflated Sharpe ratio asks whether the winner survives its own selection. This asks what bar the search sets in the first place, and it needs the whole trial panel rather than a summary: one column per trial that was evaluated.

### What the number is, and what it is not

Every trial is treated as null when counting expected false discoveries, and what comes out is `E[V] / R`: expected false discoveries over the realized discovery count. That is not `E[V / R]`, and the two do not agree in general. It is a Benjamini-Hochberg / Storey-style plug-in with `m0 = m`.

It is **not** the double bootstrap of Harvey and Liu (2020), which ranks t-statistics in an outer bootstrap, assigns alternatives from a supplied share of true strategies, and averages the realized `FP / (FP + TP)` in an inner bootstrap — inputs this signature does not take. Their paper is why a hurdle of this shape is worth having; the number here is not theirs. The docstring now says that in those words, and the title and framing of this PR have been corrected to match. Thanks to @timkpaine for catching the overclaim.

### There is no universal t > 3

The bar depends on how many candidates were tried and on how correlated they were, so it is derived from the search rather than read off a table. On a panel of 60 trials with no skill whatsoever (500 daily observations, seeded):

```python
hurdle = trials.calc_fdr_hurdle()          # 2.20
trials.apply(lambda c: c.mean() / (c.std() / len(c) ** 0.5)).abs().max()   # 2.18
```

The best of those 60 trials reaches |t| = 2.18, which reads as significant on its own and is nothing of the kind: the search that produced it sets a bar of 2.20, and nothing clears it.

Plant three genuinely skilled trials in the same panel and widen the search around them:

| trials evaluated | hurdle | discoveries |
|---|---|---|
| 5 | 2.20 | 3 |
| 20 | 2.60 | 4 |
| 60 | 2.95 | 3 |

The same three strategies, the same data, a rising bar. That is the whole content of the method, and it is why the trials that were discarded belong in the panel.

### How the null is built

Each column is demeaned and the time index is resampled, drawing the **same** periods for every column so that cross-trial correlation survives into the null rather than being assumed away. The hurdle returned is the smallest one for which the target also holds at every stricter hurdle, so it cannot land on a dip in a noisy curve.

Because the alternative is never represented, the function bounds false discoveries only and says nothing about missed ones. The docstring states that rather than leaving it to be inferred.

### Missing data is complete-case, and no longer silent

A period is dropped unless every trial has an observation there. That is forced by the shared draw above — the bootstrap takes one period for all columns at once, and there is nothing to take where a column has no observation — so `GroupStats`, which keeps each series on its own calendar, cannot be followed here.

Trials with staggered histories can lose most of the sample that way, and the hurdle moves with it. So the periods kept and the periods given are now reported in a warning whenever any row goes, the docstring names the trade-off, and a test asserts the returned hurdle is the one those shared periods imply rather than the one the same trials give over their whole history.

### Two-sided, by design and now in writing

Trials are ranked on `|t|`, so a strongly negative strategy is a discovery too. A caller screening for outperformance should compare the hurdle against the positive-t columns only. Documented, and covered by a test that plants a negative trial: it clears the hurdle on `|t|` while nothing clears it on `t`.

### Zero-dispersion guard, from the start

A constant column has no t-statistic, but it does not arrive as a `nan`: the standard deviation of a constant series is floating-point residue rather than an exact zero, so a flat column divides out to something enormous but finite and would be counted as the strongest trial in the search. Such columns are dropped, and the docstring says so, since dropping changes the trial count that sets the hurdle.

This is the same failure mode this repo fixed for the deflated Sharpe ratio in #317, so the guard ships with the feature instead of arriving as a follow-up. Its test sweeps scales 1e-12 to 1e3 across lengths 50 to 1000, rather than resting on the single point the floor was calibrated on — a genuinely low-volatility panel still gets a number, only the degenerate one does not.

### Verification

- Agreement to the grid resolution with a separate implementation of the same estimator that I maintain, on both a null panel and one with planted signal (2.20 vs 2.20, 2.85 vs 2.85).
- Seeded, so the hurdle is reproducible; equality across repeated calls asserted in the tests.
- 81/81 tests pass (7 new). `ruff check` and `ruff format` introduce nothing new: verified by running both against `master` and diffing the findings, since the repo carries some pre-existing ones.
- No new dependency beyond stdlib `warnings`; numpy and pandas only otherwise.

Happy to rename the function if `calc_fdr_hurdle` still overstates it now that the estimand is spelled out — `calc_multiplicity_hurdle` would be the honest alternative.

### CODE REVIEW

File: ffn/core.py [hunk 1]
```diff
@@ -2608,6 +2608,121 @@ def calc_deflated_sharpe_ratio(returns, trial_sharpe_ratios, rf=0.0, nperiods=No
     return scipy.stats.norm.cdf((sr - sr0) * np.sqrt(n - 1) / np.sqrt(variance_adj))
 
 
+def calc_fdr_hurdle(returns, target_fdr=0.05, n_boot=1000, seed=0, grid_step=0.05):
+    """
+    Calculates the t-statistic hurdle that a strategy search implies at a chosen
+    false discovery rate, following `Harvey and Liu (2020)
+    <https://doi.org/10.1111/jofi.12951>`_.
```
Comment (timkpaine) [inline ffn/core.py]: This does not implement the cited procedure. Harvey–Liu Steps I–IV rank t-statistics in an outer bootstrap, assign alternatives based on a user-supplied `p0`, then use an inner bootstrap to average realized `FP / (FP + TP)`. This function has no `p0`, alternatives, or nested bootstrap. What is computed here is a defensible BH/Storey-style plug-in estimate with every trial treated as null, but it is not the Harvey–Liu hurdle. Cheapest fix is to remove the “following Harvey and Liu” attribution and describe this as a bootstrap plug-in hurdle. Implementing their double bootstrap would require a different API.

File: ffn/core.py [hunk 2]
```diff
@@ -2608,6 +2608,121 @@ def calc_deflated_sharpe_ratio(returns, trial_sharpe_ratios, rf=0.0, nperiods=No
...
+            tb = np.abs(sample.mean(axis=0) / (sd / root_n))
+        null[b * m : (b + 1) * m] = np.where(np.isfinite(tb), tb, 0.0)
+    null.sort()
+
+    grid = np.arange(0.0, max(4.0, obs[-1] + 0.5) + grid_step, grid_step)
+    # Expected false discoveries at each hurdle, and the discoveries actually seen.
+    expected_false = (len(null) - np.searchsorted(null, grid, side="left")) / n_boot
+    discoveries = m - np.searchsorted(obs, grid, side="left")
+    with np.errstate(divide="ignore", invalid="ignore"):
+        fdr = np.where(discoveries > 0, np.minimum(1.0, expected_false / discoveries), 0.0)
```
Comment (timkpaine) [inline ffn/core.py:2754]: Worth noting here that this computes `E[V] / R`, not Harvey–Liu `E[V/R]`; they are not generally equivalent.

File: ffn/core.py [hunk 3]
```diff
@@ -2608,6 +2608,121 @@ def calc_deflated_sharpe_ratio(returns, trial_sharpe_ratios, rf=0.0, nperiods=No
...
+
+    """
+    if not 0.0 < target_fdr < 1.0:
+        raise ValueError("target_fdr must be between 0 and 1")
+    if n_boot < 100:
+        raise ValueError("n_boot below 100 makes the null too coarse to trust")
+    if grid_step <= 0:
+        raise ValueError("grid_step must be positive")
+
+    x = pd.DataFrame(returns).dropna(axis=0, how="any")
```
Comment (timkpaine) [inline ffn/core.py]: `dropna(how="any")` removes a period from every trial whenever one trial is missing it. Wide panels commonly have staggered histories, so this can silently discard most observations and materially change the hurdle. Existing `GroupStats` code deliberately avoids dropping rows from unrelated series this way. Could this require a balanced panel explicitly, or clearly document complete-case deletion and effective sample size?

File: ffn/core.py [hunk 4]
```diff
@@ -2608,6 +2608,121 @@ def calc_deflated_sharpe_ratio(returns, trial_sharpe_ratios, rf=0.0, nperiods=No
...
+    if not keep.any():
+        return np.nan
+    dropped = m - int(keep.sum())
+    values = values[:, keep]
+    m = values.shape[1]
+    if m < 2:
+        raise ValueError(f"only {m} of {m + dropped} trials carry dispersion; need 2")
+
+    root_n = np.sqrt(n)
+    obs = np.sort(np.abs(values.mean(axis=0) / (values.std(axis=0, ddof=1) / root_n)))
```
Comment (timkpaine) [inline ffn/core.py:2734]: Please document that this is intentionally two-sided. Harvey–Liu main application is one-sided, although they discuss two-sided testing for data-mined panels. With `abs(t)`, a strongly negative strategy is also a discovery, which callers screening for outperformance may not expect.

**Review by timkpaine — CHANGES_REQUESTED** [review]
The statistical attribution and missing-data behavior need resolution before this can merge. Relabeling the estimator is a valid alternative to implementing the full Harvey–Liu procedure; the two-sided behavior only needs explicit documentation.

File: ffn/core.py (hunk 1 above)
Comment (ipezygj) [reply ffn/core.py]: Corrected. You are right, and it should not have shipped that way.

The docstring now says what the function computes — a bootstrap plug-in, every trial treated as null in the numerator, conservative in the same way Benjamini-Hochberg is with `m0 = m` — and in the same paragraph states that this is *not* the Harvey-Liu procedure, naming the outer bootstrap, the supplied share of true strategies, and the inner bootstrap as the things it does not have and could not take through this signature. Harvey and Liu are kept only as the reason a hurdle of this shape is worth having: "Their paper is why a hurdle of this shape is worth having; the number here is not theirs."

The PR title and body are corrected the same way, since the overclaim was in both.

One thing I did not decide unilaterally: if `calc_fdr_hurdle` still overstates it now that the estimand is written down, `calc_multiplicity_hurdle` is the honest alternative and I will rename before merge. Your call.

File: ffn/core.py (hunk 2 above)
Comment (ipezygj) [reply ffn/core.py:2754]: Stated, in both places.

The docstring: "What this estimates is `E[V] / R`, expected false discoveries over the realized discovery count. That is not `E[V / R]`, and the two do not agree in general."

And the comment on the lines you flagged, so it sits next to the division that makes it:

```python
    # Expected false discoveries at each hurdle, and the discoveries actually seen. Their
    # ratio is E[V] / R, which is not the same quantity as E[V / R].
```

File: ffn/core.py (hunk 3 above)
Comment (ipezygj) [reply ffn/core.py]: Kept, but no longer quiet — and now argued for instead of assumed.

The reason it cannot follow `GroupStats` is the line right above it: the bootstrap draws one period for **every** column at once, and that shared draw is exactly what keeps cross-trial correlation in the null instead of assuming it away. Where a column has no observation at that period there is nothing to draw. So the balanced panel is forced by how the null is built, not by convenience. That is now written where the `dropna` happens, and in the docstring.

What was actually wrong is that it happened silently. It now warns with both numbers:

```
calc_fdr_hurdle kept 100 of 400 periods: a period is dropped unless all 20 trials have an
observation there. Trials with staggered histories lose most of the sample this way, and
the hurdle is the one those periods imply.
```

and the docstring names the trade-off and points the caller at aligning the panel themselves if they would rather make it explicitly.

You are right that it moves the hurdle, so the new test measures how far. Twenty trials over 400 periods, two of them with a real edge, one trial starting 300 periods late:

| panel | periods used | hurdle | discoveries |
|---|---|---|---|
| as given | 400 | 2.60 | 3 |
| after complete-case deletion | 100 | 2.90 | 2 |

One in four rows survives, the bar rises, and a trial that is a discovery on the full sample stops being one. The test asserts the warning and its count, that the returned value is the one those 100 shared periods imply, and that the discovery count is the smaller of the two — so the consequence is pinned, not just the mechanism.

I stopped short of raising instead of warning: any "too much was dropped" rule needs a threshold I cannot justify, and a hard-coded one would be a number that quietly moves the answer. If you would rather it raise unconditionally on any NaN and make the caller align first, that is a two-line change and I will make it.

File: ffn/core.py (hunk 4 above)
Comment (ipezygj) [reply ffn/core.py:2734]: Documented, as its own paragraph rather than a clause:

> The comparison is two-sided: trials are ranked by `|t|`, so a strongly negative strategy counts as a discovery too. A caller screening for outperformance should compare the hurdle against the positive-t columns only.

Also pinned by a test rather than only described — it plants a trial with a genuine negative edge and asserts it clears the hurdle on `|t|` while nothing clears it on `t`. If the behaviour ever changes to one-sided, that test says so.
